### DISCUSSION TITLE
Do we still need "Update site project" wizard and related editors in PDE

### DISCUSSION THREAD

**Discussion by vogella** [eclipse-pde/eclipse.pde#113] [General]
AFAIK site project is not needed anymore for creating update sites. A project of type General with category.xml is sufficient. Do we still have a use base for this, or shall we retire the related code?

cc @laeubi @HannesWell

**Comment by merks** [comment]
When you ask "Do **we** still need...", which **we** do you mean?  The use base is the community at large so asking two people this question is a pretty narrow survey result...

**Reply by mickaelistria** [reply]
I think it's indeed not the right question: the question would be more "Do we still want to offer -and maintain- this functionality to the community despite X, Y, Z...?". Here it seems clear that no-one has real need for it as there has been a recommended better technical alternative for more than a decade. The need is really non-existent, in worst case, it would be a false impression if someone say they need it because they never considered doing differently.
IMO, the committers have for duty to *lead* the users in being more productive and achieving higher quality; not the duty to eternally maintain workflows that are less productive. So even if many users do use it, it can still be a good thing to remove it, it will serve those users better on the long run.

**Reply by vogella** [reply]
@mickaelistria I just wanted to see if update sites still provides something which category does not before starting the remove it discussion. But of course I agree with your point that we should retire oudated stuff. This will help users to use the latest and greatest and to reduce our legacy code base.

**Comment by vogella** [comment]
@merks we as in active PDE develpers and people watching this repo.

So are we (including @merks and any PDE developer and user who is watching this repo) aware of a use case which requires this project type?

**Comment by laeubi** [comment]
By the way you can also create a "poll" discussion... anyways I never used that project type actively and tycho-pomless is just looking at the `category.xml`.

**Comment by merks** [comment]
@mickaelistria  That is indeed a better way to phrase it.   Certainly we can strongly argue that there should be no wizard because we should not encourage creating such thing.   If someone is still using it (and p2 does still support loading such a site), they can always edit via an XML editor so no really compelling need for a dedicated editor.  So personally I think it would be okay to get rid of it.  But that's just my opinion; I think removing p2's support for it would be unhelpful, but that's not the discussion here.

Note that I see these site.xml resources in the SDK workspace:

![image](https://user-images.githubusercontent.com/208716/168764972-bbe69554-dcfd-490d-a11f-d0e743db13d3.png)

**Comment by vogella** [comment]
Nice, nice. Created https://github.com/eclipse-pde/eclipse.pde/issues/114 for the actual removal now that "we" :-) all agree.

**Reply by laeubi** [reply]
Maybe the editor can be replaced by something that ask to convert a `site.xml` -> `category.xml` e.g. in most case you can simply rename the file and you are done... don't know if there are any specialties I'm not aware of but that's how I usually "convert" such old ~~code~~ sites.

**Comment by SarikaSinha** [comment]
I just used this wizard a month back to create an update site project to add a "marketplace entry".
This is definitely very easy and convenient way to to host our changes.

**Comment by vogella** [comment]
@SarikaSinha sounds like you one of the users we should guide to use better technologies. ;-) Joking aside, did you try to use the category.xml and faced any issue with it?

**Reply by iloveeclipse** [reply]
> A simple project with category.xml is sufficient.

May be, but it is also discoverable by users? So if a user thinks "I would like to create update site" and goes to the "New" wizard, will it be able to finish his task without "new update site wizard"?

> @SarikaSinha sounds like you one of the users we should guide to use better technologies. ;-)

I think you've nailed it down: **we should guide**. That is exact why wizard was invented. Or should we ship platform committers bundled to SDK (instead of the wizard), so they could guide users?

> Did you try to use the category.xml and faced any issue with it?

Once again: how a user should discover that? If that is what the wizard does, so why should we remove the wizard? If the wizard doesn't offer that possibility, it should be improved.

But just removing something without providing an alternative it nonsense.

**Comment by mickaelistria** [comment]
> I think you've nailed it down: we should guide.

Keeping outdated stuff unfortunately guides, but towards wrong directions. Removing is also part of the guidance: it does remove some wrong directions.

> Once again: how a user should discover that?

I created https://github.com/eclipse-pde/eclipse.pde/pull/115 to remediate that

**Comment by jarthana** [comment]
Not specific to this issue, but it is alarming to see comments like, "never considered doings things differently". IMO, it's not our right or responsibility to force someone to do it our way especially when there are options.

**Reply by akurtakov** [reply]
It is our right to decide whether we want and whether we have the time to maintain these options.

**Reply by mickaelistria** [reply]
It's our right and our responsibility to improve usability and efficiency of the software we develop. Providing tons of options for the same goal is often a UX antipattern as it increases confusion and inconsistency, make it harder for users to take a decision, and to ensure they take the best one.
Don't get me wrong, "never considered doings things differently" is not meant to be interpreted as a negative judgement towards individuals. Not immediately trying to do things differently from what we're used is part of the human nature; but it doesn't mean it's always what's best and a software needs to play with this part of "human nature" to consider how it is capable of delivering new value.
I personally find it more alarming that people who've been working daily with Eclipse IDE still want to use this wizard while the "site.xml" format was kind of deprecated more than 10 years ago. And the reason why it's still used is... because it's still there. So the cause of people not doing things differently is more a showcase that the software we write really fails at leading users to best productivity here, it basically makes some users fail at getting some of the value we create, and hence diminish the real value created (because the value created is proportional to number of users).

**Reply by laeubi** [reply]
I think we need to distinguish a bit here:

1. site.xml is long time discouraged and as far as I know there is no development in that area to improve/add features here and category.xml is a 1:1 replacement
2. as long as we support it (and don't help people to migrate) it will simply be used. One step here is that tycho no longer support building such site.xml projects, just because it is no longer maintained and require to keep old and duplicate code
3. Independent from that, we could decide if one actually needs a **dedicated** project type to host something the user name "update-site" (most users don't care if the file is named `category.xm`l or `site.xml`) or if we can use a *generic* project type, actually you can use *any* project type, goto `new > file > category definition` so the question is what do we gain with a special project type, so what do we offer here?

**Reply by iloveeclipse** [reply]
> I think we need to distinguish a bit here:

Thanks.

> 1. site.xml is long time discouraged and as far as I know there is no development in that area to improve/add features here and category.xml is a 1:1 replacement

Not really 1:1, see below.

> 2. as long as we support it (and don't help people to migrate) it will simply be used. One step here is that tycho no longer support building such site.xml projects, just because it is no longer maintained and require to keep old and duplicate code

Existing users that don't use tycho don't care, so still should be able to use site.xml.

> 3. Independent from that, we could decide if one actually needs a **dedicated** project type to host something the user name "update-site" (most users don't care if the file is named `category.xm`l or `site.xml`) or if we can use a _generic_ project type, actually you can use _any_ project type, goto `new > file > category definition` so the question is what do we gain with a special project type, so what do we offer here?

I personally like to have dedicated **project** (not project type) for update site.

But to the 1:1 story. If anyone still wants to remove site.xml from IDE and replace it with category.xml, **it should provide comparable user experience**. Today this is not the case.

Today we can create & manage simple update sites in few clicks from zero to the deployment, with only the update site editor, without using tycho/maven at all. Update site editor provides all the functionality required, and it is handy to manage simple company/team internal update sites, where users aren't maven/tycho/p2 gurus, don't use anything advanced but just want to quickly hack some internal code and give the team the updated bits.

Replacing it with category.xml would mean, they have to add maven/tycho config files, extra command line build step, learn all the things they never needed just to perform exact same tasks like before by changing version of features used & clicking in "build" button in the update site. Why? Just because we don't want to maintain the code that doesn't require any maintenance? 

I would understand this request if we would have constant flow of bug reports related to the functionality or usability of the update sites, but I can't see that. It's just of the sort "I personally don't have use for it, so let's remove it".

So **I personally** use update site, and I'm happy with it, and I don't want that IDE will loose this functionality because **I personally** would miss that and this would make **my personal life** harder.

With that, can we close this issue as "won't fix" please?

We see we have **at least one** user that needs the site.xml and we see **zero maintenance** costs to keep that.

**Reply by SarikaSinha** [reply]
At least 2, Can't agree more to this. 

Without any maintenance cost, this code has been running fine for more than 10 years. Provides a simple way to to build site without extra configuration for the quick start. Very easy to use for initial prototypes and no compelling reason to remove it.

**Reply by laeubi** [reply]
> Not really 1:1, see below.

Well from P2 POV it is 1:1 as you can do anything with `category.xml` (probably even more) as you can do with `site.xml` 

> Existing users that don't use tycho don't care

`category.xml` is not tycho specific, but tycho is the most common way to automatically build such sites.

> I personally like to have dedicated **project** (not project type) for update site.

You can create one, but it is not mandatory, you can even name it "site.xml" if it is not an update-site project it will be handled by the category editor :-)

> Replacing it with category.xml would mean, they have to add maven/tycho config files,

This is not mandatory, as mentioned above `category.xml` is not Tycho it is P2, there is just no button / export wizard to do so in PDE I think there was a bug about it but I can't found it.

**Reply by iloveeclipse** [reply]
> Well from P2 POV it is 1:1 as you can do anything with `category.xml` (probably even more) as you can do with `site.xml`

Well, from P2 point of view we don't need **any** wizard, not even an IDE.

> > Replacing it with category.xml would mean, they have to add maven/tycho config files,
> 
> This is not mandatory, as mentioned above `category.xml` is not Tycho it is P2, there is just no button / export wizard to do so in PDE I think there was a bug about it but I can't found it.

So far all what I hear was: "I don't use it", "p2 don't needs that", "go there, do that, then do next, then there is a bug to fix".

Means: today there is **no comparable alternative** that guides user and provides same functionality as site.xml editor + new update site wizard. "But you can do that and that" doesn't count, because it is not as easy and convenient.

If someone wants get rid of site.xml and new update site wizard, **please** feel free open tasks & contribute to have same or better experience with categories.xml. Once that is all done, let continue with removal of the site.xml editor & the wizard.

In general: if someone doesn't need some functionality or uses different technology for different tasks, it doesn't mean we can just remove that functionality.

* I don't use "Themes", so let remove that? 
* I don't care about "Dark mode, my system doesn't support that", so let's remove that? 
* I'm using Windows, so let remove Linux support, it causes only maintenance costs!
* I use IDE for C++ only, so let's remove JDT, it is too complex and impossible to maintain?
* I can do everything with command line, tycho, p2 & vim, so let's remove IDE code, it is too bloated!

What is this for an attitude? Is there a **need** to remove something that people use and offer them instead something that they don't want use? Just wasting our all time in such discussions, sorry.

**Reply by laeubi** [reply]
> Well, from P2 point of view we don't need **any** wizard, not even an IDE.

Correct, that's why I said before we clearly need to separate those

- from P2/Tycho (no idea about PDE build) there is not much of a difference when *building* such a site.
- from *consuming* such a produced site, as @merks has pointed out `site.xml` is bad as it produces more traffic, maybe one could fix this but no one yet has decided to do so.
- from *user* POV `category.xml` is missing a convenient method to export right from the IDE, this could be fixed, but no one yet has decided to do so.
- Then there are *editors* for both, and I don't see much difference (as a user) between both
- last but not east there is a special *project type* for site.xml that requires a wizard to be created
- then as i never used site projects I can only speak for *me* that I have no use for this and thus can't really say more than I won't miss it, but this might be different for *others* for sure.

So many different aspects would require different solutions (if there is actually a problem). And I think its always good to ask from time to time if maybe things have changed, so untill now I think the answer to "Do we still need "Update site project" wizard and related editors in PDE" seems to be "yes, unless we improve support for category.xml in the IDE" at least as today.

**Comment by merks** [comment]
Just to ensure we keep maintenance effort in perspective...  

The vast majority of the activities in this folder over the last years are nice-to-have-but-totally-unnecessary stylistic changes:

![image](https://user-images.githubusercontent.com/208716/168794158-6a4083ba-e200-44fc-92f6-b1d0caea74f9.png)

So I think it's fair to say the maintenance effort here is minimal and mostly self-inflicted...

**Comment by akurtakov** [comment]
Just as info for people considering to continue using site.xml - next Tycho version will no longer support it https://github.com/eclipse/tycho/commit/29631df290306503327d825ba9b418a633f154ba .

**Comment by trancexpress** [comment]
> AFAIK site project is not needed anymore for creating update sites. A simple project with category.xml is sufficient.

How do I create this "simple project" from Eclipse?

**Reply by laeubi** [reply]
`File > New > Project > General > Project`

**Reply by vogella** [reply]
![image](https://user-images.githubusercontent.com/139910/168818068-5dde94d2-995b-4153-9547-140466e412bb.png)

**Reply by trancexpress** [reply]
> ![image](https://user-images.githubusercontent.com/139910/168818068-5dde94d2-995b-4153-9547-140466e412bb.png)

And how do I fill the project with content? **Is this a serious answer?** The update site project wizard creates an update site project, that lets me add plug-ins etc. via ```site.xml```?

How do I create this "category.xml" and open it in a specific editor to fill it with content?

**Reply by laeubi** [reply]
As noted above `new > other > category definition` (you can do thins in any project type, under the Plugin development group) then simply double click to open the editor.

**Reply by trancexpress** [reply]
> As noted above `new > other > category definition` (you can do thins in any project type, under the Plugin development group) then simply double click to open the editor.

Thanks, how do I build from that definition?

**Reply by merks** [reply]
The category editor is quite nice:

![image](https://user-images.githubusercontent.com/208716/168821663-5054a87f-685b-45a9-a1ce-9a4874b1f821.png)

Often one doesn't (need to) create a dedicated project as in this example.  But you can if you want as in this example:

![image](https://user-images.githubusercontent.com/208716/168821982-a536caa0-292b-4c9d-940c-67d9c94c01a9.png)

From a consumer point if view, a category.xml creates a **much better quality** update site.  With a site.xml each client has to download the feature jars to get information to display to the user and that's expensive and is harsh for the server...

**Reply by iloveeclipse** [reply]
> The category editor is quite nice:

Well, site.xml editor is also quite nice, but it also offers much more functionality which is simply missing in category editor
 
![image](https://user-images.githubusercontent.com/964108/168828230-9ff6ff93-1066-4540-a154-59f058ff0c28.png)

> From a consumer point if view, a category.xml creates a **much better quality** update site. With a site.xml each client has to download the feature jars to get information to display to the user and that's expensive and is harsh for the server...

Honestly, who cares about quality of the update site if that is located on internal server or has a single bundle/feature in it? But again, I'm not against category.xml, it is just not 1:1 replacement. I'm all for removing **duplicated** functionality once we have a better choice, but we aren't there yet with category editor. And still the wizard will be needed to guide people that never created an update site in the process.

**Comment by HannesWell** [comment]
I support the intention to remove duplication from PDE.
Even if the Update-Site code did not cause maintenance effort due to explicit work on it its presence alone can cause more work because a developer has to understand the core first. So even if there are no changes it can cause 'costs'.

Furthermore it makes it complicated for the user if there are two things that actually do very similar things and I agree that we, who know the details about the technical differences, should guide the user to the best choice.

Nevertheless I also understand and agree that if we would just remove the Update-Site we would loose used functionality as it was made clear. Therefore I suggest we first enhance the `category.xml` support to reach parity and then remove the update-site support.

From the discussion I derive that there are actually two points missing in the category.xml support:
- Possibility to build the repo/site from the IDE  
  - Would it be possible to simply build a category.xml file using the current Update-Site build in the IDE? The created update-site would not be as good as with Tycho, but as @iloveeclipse said (but we could improve that later by using P2 instead):
  > Honestly, who cares about quality of the update site if that is located on internal server or has a single bundle/feature in it?
- Wizard to create a "Repository project (formerly Update-Site project)"
  - If we really want to keep the wizard (maybe for a transition period), we could just create a general project with a `category.xml` file in it.
  - Besides that I noticed that the wizard can generate a html website. But when I tried to open it in a browser it did not work.
  Is this something we want to keep?

**Comment by merks** [comment]
A site.xml seems to only support Add Feature... while a category.xml includes Add Bundle...  so it seems one would not generally be able to map a more flexible category.xml to a more restrictive site.xml.
